## Author: Rami Bazoqa
## Date: 4/4/2020
## CS301-006, Professor Watson
## HW06 Solution

The purpose of this assignment is to help students practice using Python to prepare their data sets for machine learning.

link to the git repo: https://github.com/Bazoqa/CS301_HW

In [252]:
import pandas as pd
import numpy as np

#### Question 1 - Read in the data set and print the % empty values for each column. Are there any columns with missing values?

In [253]:
data_dir = 'C:\\Users\\Rami\\Downloads\\groundhog.csv'

In [254]:
df = pd.read_csv(data_dir)
df = df.rename(columns={"Unnamed: 0": "rowID"})
df

rowID  Year Punxsutawney Phil  February Average Temperature  \
0        0  1886         No Record                           NaN   
1        1  1887       Full Shadow                           NaN   
2        2  1888       Full Shadow                           NaN   
3        3  1889         No Record                           NaN   
4        4  1890         No Shadow                           NaN   
..     ...   ...               ...                           ...   
126    126  2012       Full Shadow                         37.51   
127    127  2013         No Shadow                         34.77   
128    128  2014       Full Shadow                         32.13   
129    129  2015       Full Shadow                         32.99   
130    130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                            NaN                        NaN   
1                                            NaN                        NaN   
2                                            NaN                        NaN   
3                                            NaN                        NaN   
4                                            NaN                        NaN   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                                 41.6   
130                                 49.5   

   

In [255]:
df.isnull().sum() * 100 / len(df)

rowID                                          0.000000
Year                                           0.000000
Punxsutawney Phil                              0.000000
February Average Temperature                   6.870229
February Average Temperature (Northeast)       6.870229
February Average Temperature (Midwest)         6.870229
February Average Temperature (Pennsylvania)    6.870229
March Average Temperature                      6.870229
March Average Temperature (Northeast)          6.870229
March Average Temperature (Midwest)            6.870229
March Average Temperature (Pennsylvania)       6.870229
dtype: float64

Yes, columns named "February Average Temperature" through "March Average Temperature (Pennsylvania)" are missing 9 entries of data (~6.87%)

#### Question 2 - Create a new dataframe which is a copy of the orginal one. In this copy, replace the missing values in the data set using one of the discussed methods in class.

In [256]:
df_copy = df.copy()

In [257]:
df_copy["February Average Temperature"].fillna(df_copy["February Average Temperature"].mean(),inplace=True)
df_copy["February Average Temperature (Northeast)"].fillna(df_copy["February Average Temperature (Northeast)"].mean(),inplace=True)
df_copy["February Average Temperature (Midwest)"].fillna(df_copy["February Average Temperature (Midwest)"].mean(),inplace=True)
df_copy["February Average Temperature (Pennsylvania)"].fillna(df_copy["February Average Temperature (Pennsylvania)"].mean(),inplace=True)
df_copy["March Average Temperature"].fillna(df_copy["March Average Temperature"].mean(),inplace=True)
df_copy["March Average Temperature (Northeast)"].fillna(df_copy["March Average Temperature (Northeast)"].mean(),inplace=True)
df_copy["March Average Temperature (Midwest)"].fillna(df_copy["March Average Temperature (Midwest)"].mean(),inplace=True)
df_copy["March Average Temperature (Pennsylvania)"].fillna(df_copy["March Average Temperature (Pennsylvania)"].mean(),inplace=True)
df_copy.head()

rowID  Year Punxsutawney Phil  February Average Temperature  \
0      0  1886         No Record                     33.802705   
1      1  1887       Full Shadow                     33.802705   
2      2  1888       Full Shadow                     33.802705   
3      3  1889         No Record                     33.802705   
4      4  1890         No Shadow                     33.802705   

   February Average Temperature (Northeast)  \
0                                 22.693443   
1                                 22.693443   
2                                 22.693443   
3                                 22.693443   
4                                 22.693443   

   February Average Temperature (Midwest)  \
0                               32.694262   
1                               32.694262   
2                               32.694262   
3                               32.694262   
4                               32.694262   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                    26.522951                  41.699016   
1                                    26.522951                  41.699016   
2                                    26.522951                  41.699016   
3                                    26.522951                  41.699016   
4                                    26.522951                  41.699016   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                              32.368033                            42.568852   
1                              32.368033                            42.568852   
2                              32.368033                            42.568852   
3                              32.368033                            42.568852   
4                              32.368033                            42.568852   

   March Average Temperature (Pennsylvania)  
0                                 35.908197  
1                                 35.908197  
2                                 35.908197  
3                                 35.908197  
4                                 35.908197

#### Question 3 - Is your proposed method from Question 2 appropriate in this case? Why or why not?

Yes, I replaced missing values with the mean for the column, which is appropriate for Numeric features (which temperatures are)

#### Question 4 - Using the original dataframe (unchanged from Questoins 2, 3), create a mapping for the "Punxsutawney Phil" column which changes the categories into numerical quantities. For example: 'No Record' can be 0, 'No Shadow' can be 1, 'Partial Shadow' can be 2, and 'Full Shadow' can be 3. You should create the mapping *and* apply it to the data set by adding a new column called "Shadow".

In [258]:
df["Shadow"] = -1
df.loc[df["Punxsutawney Phil"] == "No Record", "Shadow"] = 0
df.loc[df["Punxsutawney Phil"] == "No Shadow", "Shadow"] = 1
df.loc[df["Punxsutawney Phil"] == "Partial Shadow", "Shadow"] = 2
df.loc[df["Punxsutawney Phil"] == "Full Shadow", "Shadow"] = 3
df.head(10)

rowID  Year Punxsutawney Phil  February Average Temperature  \
0      0  1886         No Record                           NaN   
1      1  1887       Full Shadow                           NaN   
2      2  1888       Full Shadow                           NaN   
3      3  1889         No Record                           NaN   
4      4  1890         No Shadow                           NaN   
5      5  1891         No Record                           NaN   
6      6  1892         No Record                           NaN   
7      7  1893         No Record                           NaN   
8      8  1894         No Record                           NaN   
9      9  1895         No Record                          26.6   

   February Average Temperature (Northeast)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
5                                       NaN   
6                                       NaN   
7                                       NaN   
8                                       NaN   
9                                      15.6   

   February Average Temperature (Midwest)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
5                                     NaN   
6                                     NaN   
7                                     NaN   
8                                     NaN   
9                                    21.9   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          NaN                        NaN   
1                                          NaN                        NaN   
2                                          NaN                        NaN   
3                                          NaN                        NaN   
4                                          NaN                        NaN   
5                                          NaN                        NaN   
6                                          NaN                        NaN   
7                                          NaN                        NaN   
8                                          NaN                        NaN   
9                                         17.0                      39.97   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                    NaN                                  NaN   
4                                    NaN                                  NaN   
5                                    NaN                                  NaN   
6                                    NaN                                  NaN   
7                                    NaN                                  NaN   
8                                    NaN                                  NaN   
9                                   27.6                                 40.2   

   March Average Temperature (Pennsylvania)  Shadow  
0                                       NaN       0  
1                                       NaN       3  
2                                       NaN       3  
3                                       NaN       0  
4                                       NaN       1  
5                                       NaN       0  
6                                       NaN       0  
7                                       NaN       0  
8                                   

#### Question 5 - For all rows which have no record, remove the rows which have empty values in all of the quantitative columns. (Only 6 should be removed.)

In [259]:
df_nulls = df.copy()
df_nulls = df_nulls[df_nulls.isnull().any(axis=1)]
df_nulls = df_nulls.loc[df_nulls["Shadow"] > 0]

df_full = df.copy()
df_full.dropna(how="all", inplace=True, subset=['February Average Temperature', 'February Average Temperature (Northeast)', 'February Average Temperature (Midwest)', 'February Average Temperature (Pennsylvania)', 'March Average Temperature', 'March Average Temperature (Northeast)', 'March Average Temperature (Midwest)', 'March Average Temperature (Pennsylvania)'])

df_1 = pd.concat([df_nulls, df_full], ignore_index="True")

df = df_1.copy()
df.head(10)

rowID  Year Punxsutawney Phil  February Average Temperature  \
0      1  1887       Full Shadow                           NaN   
1      2  1888       Full Shadow                           NaN   
2      4  1890         No Shadow                           NaN   
3      9  1895         No Record                         26.60   
4     10  1896         No Record                         35.04   
5     11  1897         No Record                         33.39   
6     12  1898       Full Shadow                         35.37   
7     13  1899         No Record                         25.50   
8     14  1900       Full Shadow                         30.76   
9     15  1901       Full Shadow                         29.86   

   February Average Temperature (Northeast)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                      15.6   
4                                      22.2   
5                                      23.6   
6                                      24.8   
7                                      18.1   
8                                      21.4   
9                                      16.0   

   February Average Temperature (Midwest)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                    21.9   
4                                    33.5   
5                                    34.7   
6                                    33.3   
7                                    22.2   
8                                    27.5   
9                                    26.2   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          NaN                        NaN   
1                                          NaN                        NaN   
2                                          NaN                        NaN   
3                                         17.0                      39.97   
4                                         26.6                      38.03   
5                                         27.9                      38.79   
6                                         26.7                      41.05   
7                                         20.0                      37.63   
8                                         24.1                      41.27   
9                                         18.9                      40.71   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                   27.6                                 40.2   
4                                   25.3                                 36.9   
5                                   32.0                                 44.0   
6                                   38.0                                 46.0   
7                                   29.3                                 38.4   
8                                   26.0                                 37.9   
9                                   30.8                                 41.9   

   March Average Temperature (Pennsylvania)  Shadow  
0                                       NaN       3  
1                                       NaN       3  
2                                       NaN       1  
3                                      31.3       0  
4                                      27.8       0  
5                                      36.9       0  
6                                      42.0       3  
7                                      34.0       0  
8                                   

#### Question 6 - For the remaining rows which have no record (but the other columns *are* populated), determine an appropriate method for inputting the value for this column. Explain your reasoning, and show your work below (using pandas).

In [260]:
print(df.loc[df["Punxsutawney Phil"] == "No Record", "February Average Temperature"].mean())
print(df.loc[df["Punxsutawney Phil"] == "Full Shadow", "February Average Temperature"].mean())
print(df.loc[df["Punxsutawney Phil"] == "Partial Shadow", "February Average Temperature"].mean())
print(df.loc[df["Punxsutawney Phil"] == "No Shadow", "February Average Temperature"].mean())

31.378333333333334
33.7125
30.74
35.577999999999996


##### Explanation

Since The mean for the remaining "NO Record" rows is closest to the mean for "Partial Shadow" rows, we will fill in the "No Record" rows with "Partial Shadow"

In [261]:
df.loc[df["Punxsutawney Phil"] == "No Record", "Punxsutawney Phil"] = "Partial Shadow"
df.head(10)

rowID  Year Punxsutawney Phil  February Average Temperature  \
0      1  1887       Full Shadow                           NaN   
1      2  1888       Full Shadow                           NaN   
2      4  1890         No Shadow                           NaN   
3      9  1895    Partial Shadow                         26.60   
4     10  1896    Partial Shadow                         35.04   
5     11  1897    Partial Shadow                         33.39   
6     12  1898       Full Shadow                         35.37   
7     13  1899    Partial Shadow                         25.50   
8     14  1900       Full Shadow                         30.76   
9     15  1901       Full Shadow                         29.86   

   February Average Temperature (Northeast)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                      15.6   
4                                      22.2   
5                                      23.6   
6                                      24.8   
7                                      18.1   
8                                      21.4   
9                                      16.0   

   February Average Temperature (Midwest)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                    21.9   
4                                    33.5   
5                                    34.7   
6                                    33.3   
7                                    22.2   
8                                    27.5   
9                                    26.2   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          NaN                        NaN   
1                                          NaN                        NaN   
2                                          NaN                        NaN   
3                                         17.0                      39.97   
4                                         26.6                      38.03   
5                                         27.9                      38.79   
6                                         26.7                      41.05   
7                                         20.0                      37.63   
8                                         24.1                      41.27   
9                                         18.9                      40.71   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                   27.6                                 40.2   
4                                   25.3                                 36.9   
5                                   32.0                                 44.0   
6                                   38.0                                 46.0   
7                                   29.3                                 38.4   
8                                   26.0                                 37.9   
9                                   30.8                                 41.9   

   March Average Temperature (Pennsylvania)  Shadow  
0                                       NaN       3  
1                                       NaN       3  
2                                       NaN       1  
3                                      31.3       0  
4                                      27.8       0  
5                                      36.9       0  
6                                      42.0       3  
7                                      34.0       0  
8                                   

#### Question 7 - If we are trying to write a machine learning program which predicts the type of shadow for the year, which column(s) should be the independent variables X? Why? 

#### Use pandas to assign a variable X to be all of the rows of the data set with the column(s) that represent the independent variables.

The independent variables should be the 8 data columns that have recorded temperatures, because this data may allow us to predict the type of shadow based on past results

In [276]:
X = df.copy()

In [277]:
X = X.drop("rowID",1)
X = X.drop("Year",1)
X = X.drop("Punxsutawney Phil",1)
X = X.drop("Shadow",1)
X.head()

February Average Temperature  February Average Temperature (Northeast)  \
0                           NaN                                       NaN   
1                           NaN                                       NaN   
2                           NaN                                       NaN   
3                         26.60                                      15.6   
4                         35.04                                      22.2   

   February Average Temperature (Midwest)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                    21.9   
4                                    33.5   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          NaN                        NaN   
1                                          NaN                        NaN   
2                                          NaN                        NaN   
3                                         17.0                      39.97   
4                                         26.6                      38.03   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                   27.6                                 40.2   
4                                   25.3                                 36.9   

   March Average Temperature (Pennsylvania)  
0                                       NaN  
1                                       NaN  
2                                       NaN  
3                                      31.3  
4                                      27.8

#### Question 8 - If we are trying to write a machine learning program which predicts the type of shadow for the year, which column(s) should be the dependent variables Y? Why?  

#### Use pandas to assign a variable Y to be all of the rows of the data set with the column(s) that represent the dependent variables.

Y should the the column Shadow, which is a numercial representation of the state of that year's resulting "Punxsutawney Phil" value.

In [288]:
Y = df.copy()

In [289]:
Y = Y["Shadow"]
Y.to_frame()

Shadow
0         3
1         3
2         1
3         0
4         0
..      ...
120       3
121       1
122       3
123       3
124       1

[125 rows x 1 columns]

#### Question 9 - Using the train_test_split function from the sklearn module, use a 70/30 split validation to split the data set into train and test. If you follow the example at the following link: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

Note: It's common practice to use the following variable names for unpacking the result:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=(YOUR VALUE))

sklearn will create a proper split where the following statements hold true:
X_train is your train set independent variables
Y_train is your train set dependent variables
X_test is your test set independent variables
Y_test is your test set dependent variables

In [291]:
from sklearn.model_selection import train_test_split

In [295]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.70)

#### Question 10 - Should the number of rows in X_train == number of rows in Y_train? Also, should the number of rows in X_test == Y_test? Why or why not? Validate your answer with code below:

In [319]:
print(len(X_train))
print(len(Y_train))
print(len(X_train) == len(Y_train))
print()
print(len(X_test))
print(len(Y_test))
print(len(X_test) == len(Y_test))

37
37
True

88
88
True


Yes, the number of rows in X_train should equal the number of rows in Y_train and the number of rows in X_test should equal Y_test. This is because machine learning is focused on testing inputs and outputs in order to continually improve its ability to predict/achieve the correct outputs. In order to make comparisons with its input and output, the lengths of both should be the same. In this case, the number of rows with temperature readings should be the same as the numebr of resulting shadow results.